File: e31/impl/3.py
import json
import sys

def digits(b,x):
  r=[]
  while True:
    if x==0:
      return r
    r.insert(0, x % b)
    x //= b

def less( a, b ):
  if len(a) != len(b):
    return len(a)<len(b)
  return a<b

def impl( b, upper, head, nextnums ):
  sum=0
  for x in nextnums:
    cand = head + [x]
    if less(cand, upper):
      sum+=1+impl(b, upper, cand, [x, (x+1) % b] )
  return sum

def count(b,x):
  if b==2:
    return x
  return impl( b, digits(b,x), [], range(1,b) )

def solve( src ):
  s=src.split(",")
  b = int(s[0])
  x = int(s[1],b)
  y = int(s[2],b)+1
  cx,cy = [ count(b,s) for s in [x,y] ]
  return str(cy-cx)

with open(sys.argv[1], "r") as file:
  data = json.load(file)
  for case in data["test_data"]:
    actual = solve( case["src"] )
    ok = actual == case["expected"]
    print( ("ok" if ok else "**NG**"), case["src"], actual, case["expected"] )

File: e31/impl/3.rb
# frozen_string_literal: true

require "json"

def less(x,y)
  if x.size!=y.size
    x.size < y.size
  else
    (x <=> y) < 0
  end
end

def count(b,upper,head,range)
  range.sum{ |x|
    cand = head + [x]
    if less(cand, upper)
      1+count(b, upper, cand, [x, (x+1) % b ])
    else
      0
    end
  }
end

def solve( src )
  sb, sx, sy = src.split(",")
  b = sb.to_i
  x = sx.to_i(b)
  y = (sy.to_i(b)+1)
  cx,cy = if b==2
    [x,y]
  else
    [x,y].map{ |e| count(b, e.digits(b).reverse, [], (1...b)) }
  end
  (cy-cx).to_s
end

if __FILE__==$PROGRAM_NAME
  json = File.open( ARGV[0], &:read)
  data = JSON.parse( json, symbolize_names:true )
  data[:test_data].map{ |number:, src:, expected:|
    actual = solve( src )
    (expected==actual).tap{ |ok|
      p [ (ok ? "ok" : "**NG**"), number, src, expected, actual ]
    }
  }.tap{ |r| puts( r.all? ? "everything is okay" : "something wrong" ) }
end

